lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Stemmer.java

```diff
@@ -540,8 +533,8 @@ private boolean isAffixCompatible(
       if (!isPrefix && dictionary.hasFlag(append, dictionary.compoundForbid)) {
         return false;
       }
-      WordContext allowed = isPrefix ? WordContext.COMPOUND_BEGIN : WordContext.COMPOUND_END;
-      if (context != allowed && !dictionary.hasFlag(append, dictionary.compoundPermit)) {
+      if (!context.isAffixAllowedWithoutSpecialPermit(isPrefix)
```

Comment:
This check becomes a bit more complicated with the addition of a new context, so it's extracted to a method